src/socks5proxy/src/socks5client.cpp

```diff
@@ -0,0 +1,325 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "socks5client.h"
+
+#include "socks5.h"
+
+#ifdef Q_OS_WIN
+#  include <winsock2.h>
+#  include <ws2ipdef.h>
+#else
+#  include <arpa/inet.h>
+#endif
+
+#include <QHostAddress>
+
+namespace {
+
+#ifdef Q_OS_WIN
+#  define PACK(__Declaration__) \
+    __pragma(pack(push, 1)) __Declaration__ __pragma(pack(pop))
+#else
+#  define PACK(__Declaration__) __Declaration__ __attribute__((__packed__))
+#endif
+
+PACK(struct ClientGreetingPacket {
+  uint8_t version;
+  uint8_t nauth;
+});
+
+PACK(struct ServerChoicePacket {
+  uint8_t version;
+  uint8_t cauth;
+});
+
+PACK(struct ClientConnectionRequestPacket {
+  uint8_t version;
+  uint8_t cmd;
+  uint8_t rsv;
+  uint8_t atype;
+});
+
+PACK(struct ClientConnectionAddressIpv4Packet {
+  uint8_t ip_dst[4];
+  uint16_t port;
+});
+
+PACK(struct ClientConnectionAddressDomainNamePacket {
+  uint8_t cmd;
+  uint16_t port;
+});
+
+PACK(struct ClientConnectionAddressIpv6Packet {
+  uint8_t ip_dst[16];
+  uint16_t port;
+});
+
+PACK(struct ServerResponsePacket {
+  uint8_t m_version = 0x05;
+  uint8_t m_rep = 0x01;
+  uint8_t m_rsv = 0x00;
+  uint8_t m_atype = 0x01;
+  uint32_t m_ip_dst = 0x00;
+  uint16_t m_port = 0x00;
+});
+
+ServerResponsePacket createServerResponsePacket(uint8_t rep,
+                                                uint16_t port = 0x00u) {
+  return ServerResponsePacket{0x05, rep, 0x00, 0x01, 0x00, port};
+}
+
+}  // namespace
+
+Socks5Client::Socks5Client(Socks5* parent, QTcpSocket* socket, uint16_t port)
+    : QObject(parent), m_parent(parent), m_inSocket(socket), m_socksPort(port) {
+  connect(m_inSocket, &QTcpSocket::disconnected, this, &QObject::deleteLater);
```

Comment:
Just considering the risk of dangling pointers here. Would it make more sense for the `Socks5Client` to be a child of the socket? This would give a hierarchy of: `Socks5` -> `QTcpSocket` -> `Socks5Client`.

Or maybe the `Socks5Client` should take ownership of the provided socket, in which case we would get:
`Socks5` -> `Socks5Client` -> [`m_inSocket`, `m_outSocket`]

The thing we don't want is for some other parent to garbage collect the `QTcpSocket` when the `Socks5Client` is still pointing to it.